### PR TITLE
Relocate mock backends from qiskit.test.mock to qiskit.mock

### LINKED ISSUE #3905: Relocate mock backends from qiskit.test.mock to qiskit.mock or similar
### What is the expected enhancement?

From https://github.com/Qiskit/qiskit-terra/pull/3890#issuecomment-591655403 : now that the fake backends are useful beyond testing, they should be moved somewhere more visible and  their usage should be documented.

### PR DESCRIPTION
**Pull request by iuliazidaru**
### Summary
fixes #3905 Relocate mock backends from qiskit.test.mock to qiskit.mock or similar 

### Details and comments

In this implementation the backends are accessible from qiskit.mock.
The deprecation will be implemented in other PR.

### CODE REVIEW

**Comment by iuliazidaru** [comment]
@jakelishman Any news on this PR?

**Comment by jakelishman** [comment]
Sorry about leaving this for a while - I forgot to ask the team about it after the new year, so I've done that now.  We're thinking that it'll be best to put them in `qiskit.providers.<something>`, since they're really fake versions of IBM Q hardware, but there's some discussion going on internally about what the `<something>` should be - some of us (including me) aren't a huge fan of using the word "mock", because the classes aren't really purely for testing any more.

Either I or someone else can reply once we've made up our minds - I'll just tag @kdk in case I'm not around in a couple of days' time.

**Comment by kdk** [comment]
Thanks for taking this on @iuliazidaru . After some discussion, it seems like the best approach will be to relocate these to `qiskit.providers.fake_provider` so that these devices can be accessed similar to those of BasicAer.

File: test/python/mock/test_new_qasm_backend.py
Comment (jakelishman) [inline test/python/mock/test_new_qasm_backend.py:18]: The right place to get this object from is `qiskit.utils.optionals.HAS_AER` - I didn't even realise there was old stuff kicking around in `qiskit.utils.backend_utils`, but come Terra 0.21 we'll be able to mark it as deprecated and get rid of it.

File: qiskit/providers/fake_provider/__init__.py
Comment (jakelishman) [inline qiskit/providers/fake_provider/__init__.py:23]: When things are far away, it's best to reference them by complete name (e.g. `from qiskit.test.mock.fake_provider`) for legibility.

File: test/python/mock/test_new_qasm_backend.py
Comment (jakelishman) [inline test/python/mock/test_new_qasm_backend.py:36]: What are the two new tests for?

**Review by jakelishman — COMMENTED** [review]
Thanks for looking at this and sticking with it.

If you're up for it, it'd be good to replace all uses of `qiskit.test.mock` with the corresponding access through `qiskit.providers.fake_provider`, anticipating the deprecation in the future, but that's a lot of busy work and if you don't feel like it, one of the core team can take care of it when we actually deprecate the objects.

File: test/python/mock/test_new_qasm_backend.py
Comment (iuliazidaru) [reply test/python/mock/test_new_qasm_backend.py:36]: The tests are for the new imports. I didn't change old code as we need it still tested. Related to the comment above, if we replace all the usages right now, then the old code doesn't have tests anymore. I can do that, but maybe it is better to change all usages when the deprecation is introduced. If a new task is created, you can assign it to me.

**Comment by HuangJunye** [comment]
@iuliazidaru Thank you for making the contribution! @jakelishman is there anything else is needed to change in order to merge this issue? I can help review this as well if you want a second opinion.

File: test/python/mock/test_new_qasm_backend.py
Comment (jakelishman) [reply test/python/mock/test_new_qasm_backend.py]: That's fair enough.  Really, the same code is tested as before, but this does test the new import paths, which is useful.

**Review by jakelishman — DISMISSED** [review]
This looks good to me, thanks for doing it!

Once Terra 0.20 is out, we can actually move the code over and begin issuing the deprecations.

File: releasenotes/notes/access-backends-from-mock-d3897ecb8490219a.yaml
Comment (jakelishman) [inline releasenotes/notes/access-backends-from-mock-d3897ecb8490219a.yaml:3]:
```suggestion
upgrade:
```

**Review by jakelishman — APPROVED** [review]

**Pull request merged**
